src/main/java/io/aiven/kafka/connect/s3/S3OutputStream.java

```diff
@@ -17,97 +17,152 @@
+        this.byteBuffer = ByteBuffer.allocate(partSize);
     }
 
-    private void expandBuffer(final int dataLength) {
-        if (this.bufferSize - this.bufferLen < dataLength) {
-            int newBufferSize = this.bufferSize;
-            while (newBufferSize - this.bufferLen < dataLength) {
-                newBufferSize = newBufferSize * 2;
-            }
-            final byte[] newBuffer = new byte[newBufferSize];
-            System.arraycopy(this.buffer, 0, newBuffer, 0, this.bufferLen);
-            this.buffer = newBuffer;
-            this.bufferSize = newBufferSize;
-        }
+    @Override
+    public void write(final int b) throws IOException {
+        write(new byte[]{(byte) b}, 0, 1);
     }
 
     @Override
-    public void write(final byte[] data, final int offset, final int len) {
-        this.expandBuffer(len);
-        System.arraycopy(data, offset, this.buffer, this.bufferLen, len);
-        this.bufferLen += len;
+    public void write(final byte[] b, final int off, final int len) throws IOException {
+        if (Objects.isNull(b) || b.length == 0) {
+            return;
+        }
+        if (Objects.isNull(multipartUpload)) {
+            multipartUpload = newMultipartUpload();
+        }
+        final var source = ByteBuffer.wrap(b, off, len);
+        while (source.hasRemaining()) {
+            final var transferred = Math.min(byteBuffer.remaining(), source.remaining());
+            final var offset = source.arrayOffset() + source.position();
+            byteBuffer.put(source.array(), offset, transferred);
+            source.position(source.position() + transferred);
+            if (!byteBuffer.hasRemaining()) {
+                flushBuffer(0, partSize, partSize);
+            }
+        }
     }
 
-    @Override
-    public void write(final byte[] data) {
-        this.write(data, 0, data.length);
+    private MultipartUpload newMultipartUpload() throws IOException {
+        logger.debug("Create new multipart upload request");
+        final var initialRequest = new InitiateMultipartUploadRequest(bucketName, key);
+        final var initiateResult = client.initiateMultipartUpload(initialRequest);
+        logger.debug("Upload ID: {}", initiateResult.getUploadId());
+        return new MultipartUpload(initiateResult.getUploadId());
     }
 
     @Override
-    public void write(final int dataByte) {
-        this.expandBuffer(1);
-        this.buffer[this.bufferLen] = (byte) dataByte;
-        this.bufferLen += 1;
+    public void close() throws IOException {
+        if (closed) {
+            return;
+        }
+        if (byteBuffer.position() > 0 && Objects.nonNull(multipartUpload)) {
+            flushBuffer(byteBuffer.arrayOffset(), byteBuffer.position(), byteBuffer.position());
+        }
+        if (Objects.nonNull(multipartUpload)) {
+            multipartUpload.complete();
+            multipartUpload = null;
+        }
+        closed = true;
+        super.close();
```

Comment:
Seems that parent's `close` method is empty, so why this call is needed?

Reply:
well I prefer to use it this way. Lets say a best practice